app/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,228 @@
+    override val messages: Flow<Message>
+        get() = _messages
+
+    @Volatile
+    private var isAwalaSetUp = false
+    private var awalaSetupJob: Job? = null
+
+    @Volatile
+    private var isAwalaInstalledOnDevice: Boolean? = null
+
+    @Volatile
+    private var isReceivingMessages = false
+
+    init {
+        awalaSetupJob = awalaScope.launch {
+            Awala.setUp(context)
+            checkIfAwalaAppInstalled()
+            isAwalaSetUp = true
+            awalaSetupJob = null
+        }
+    }
+
+    override suspend fun sendMessage(
+        message: Message,
+        recipient: MessageRecipient,
+    ) {
+        val firstPartyEndpointNodeId = awalaRepository.getServerFirstPartyEndpointNodeId()
+            ?: registerFirstPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        val thirdPartyEndpointNodeId = when (recipient) {
+            is MessageRecipient.Server -> {
+                recipient.nodeId
+                    ?: awalaRepository.getServerThirdPartyEndpointNodeId()
+                    ?: importServerThirdPartyEndpointIfNeeded()?.nodeId
+                    ?: ""
+            }
+            else -> {
+                throw IllegalStateException("User messages are not supported yet!")
+            }
+        }
+
+        if (firstPartyEndpointNodeId.isEmpty() || thirdPartyEndpointNodeId.isEmpty()) {
+            throw IllegalStateException("some of ids is empty: $firstPartyEndpointNodeId ; $thirdPartyEndpointNodeId")
+        }
+
+        Log.d(TAG, "sendMessage() from $firstPartyEndpointNodeId to $thirdPartyEndpointNodeId: ${message})")
+
+        val firstPartyEndpoint = loadNonNullFirstPartyEndpoint(firstPartyEndpointNodeId)
+        val thirdPartyEndpoint = loadNonNullThirdPartyEndpoint(thirdPartyEndpointNodeId)
+
+        GatewayClient.sendMessage(
+            OutgoingMessage.build(
+                type = message.type.value,
+                content = message.content,
+                senderEndpoint = firstPartyEndpoint,
+                recipientEndpoint = thirdPartyEndpoint,
+            )
+        )
+    }
+
+    override suspend fun isAwalaInstalled(currentScreen: Route): Boolean {
+        if (!isAwalaSetUp) {
+            awalaSetupJob?.join()
+        }
+        return if (currentScreen == Route.AwalaNotInstalled) {
+            checkIfAwalaAppInstalled()
+        } else {
+            isAwalaInstalledOnDevice ?: checkIfAwalaAppInstalled()
+        }
+    }
+
+    private suspend fun startReceivingMessages() {
+        if (isReceivingMessages) {
+            return
+        }
+
+        awalaScope.launch {
+            Log.d(TAG, "start receiving messages...")
+            GatewayClient.receiveMessages().collect { message ->
+                // TODO Remove first message.ack() before publishing the app.
+                // It's here to avoid the server getting stuck with messages that can't be processed.
+                message.ack()
```

Comment:
BTW, that comment from the other developer is wrong: the problem he experienced was that he was using the Android Studio debugger with a breakpoint here, and the server timed out as expected. So I think we should delete it.

We should also delete the `message.ack()` call because we should only do it when the message is processed.